File: tools/py/framework_contract_checker.py
#!/usr/bin/env python3
"""
框架契约检查器

最小能力：
- 检查主文档模板是否满足 main_doc_contract.yaml 必需章节
- 检查 workflow 是否引入了未被 spec 收录的标准产物路径
- 提供框架自检模式，直接扫描当前仓库关键模板和 workflow
"""

import argparse
import json
import re
import sys
from pathlib import Path


ROOT = Path(__file__).resolve().parents[2]
CONTRACTS_DIR = ROOT / "core" / "contracts"
DEFAULT_TEMPLATE = ROOT / "templates" / "AI_Coding_Context_TEMPLATE.md"
DEFAULT_SPEC = ROOT / "core" / "framework_spec.md"
DEFAULT_WORKFLOWS = [
    ROOT / "workflows" / "generation_workflow.md",
    ROOT / "workflows" / "path_a_first_generation.md",
]
OBSOLETE_STANDARD_PATHS = {
    "dev_docs/plans/features/",
    "dev_docs/plans/bugfixes/",
}


def load_required_titles(contract_path):
    titles = []
    for line in contract_path.read_text(encoding="utf-8").splitlines():
        match = re.match(r'^\s*title:\s*"?(.*?)"?\s*$', line)
        if match:
            titles.append(match.group(1))
    return titles


def extract_h2_titles(text):
    return [match.group(1).strip() for match in re.finditer(r"^##\s+(.+?)\s*$", text, re.MULTILINE)]


def check_template(template_path, contract_path):
    required_titles = load_required_titles(contract_path)
    actual_titles = extract_h2_titles(template_path.read_text(encoding="utf-8"))
    issues = []
    for title in required_titles:
        if title not in actual_titles:
            issues.append({
                "file": str(template_path),
                "type": "missing_required_section",
                "section": title,
                "message": f"缺少必需章节: {title}",
            })
    return {"checked": 1, "issues": issues}


def extract_standard_paths(spec_path):
    text = spec_path.read_text(encoding="utf-8")
    paths = set()
    for line in text.splitlines():
        if "❌" in line:
            continue
        paths.update(re.findall(r"`(dev_docs/[^`]+)`", line))
    return paths


def normalize_standard_path(path):
    if path == "dev_docs/plans/README.md":
        return "dev_docs/plans/"
    if path.startswith("dev_docs/plans/active/"):
        return "dev_docs/plans/active/"
    if path.startswith("dev_docs/plans/done/"):
        return "dev_docs/plans/done/"
    if path.startswith("dev_docs/plans/archive/"):
        return "dev_docs/plans/archive/"
    if path.startswith("dev_docs/memos/"):
        return "dev_docs/memos/"
    if path.startswith("dev_docs/knowledge/"):
        return "dev_docs/knowledge/"
    if path.startswith("dev_docs/rules/combined/"):
        return "dev_docs/rules/combined/AI_RULES.md"
    return path


def extract_user_doc_paths(text):
    return [normalize_standard_path(match) for match in re.findall(r"dev_docs/[\w./-]+", text)]


def check_spec_paths(spec_path):
    standard_paths = sorted(extract_standard_paths(spec_path))
    issues = []
    for obsolete_path in sorted(OBSOLETE_STANDARD_PATHS):
        if obsolete_path in standard_paths:
            issues.append({
                "file": str(spec_path),
                "type": "obsolete_standard_path",
                "path": obsolete_path,
                "message": f"spec 仍包含过期标准路径: {obsolete_path}",
            })
    return {"checked": 1, "standard_paths": standard_paths, "issues": issues}


def _context_is_optional(lines, index):
    section_start = 0
    for cursor in range(index, -1, -1):
        if re.match(r"^#{2,6}\s+", lines[cursor]):
            section_start = cursor
            break
    window = lines[section_start:min(len(lines), index + 4)]
    context = "\n".join(window)
    optional_markers = ["[OPTIONAL]", "不属于", "只有在用户明确需要", "可选", "不是必需"]
    return any(marker in context for marker in optional_markers)


def check_workflow(workflow_path, spec_path):
    spec_paths = extract_standard_paths(spec_path)
    lines = workflow_path.read_text(encoding="utf-8").splitlines()
    issues = []
    seen_standard_paths = set()
    for index, line in enumerate(lines):
        for matched in extract_user_doc_paths(line):
            if matched in spec_paths:
                seen_standard_paths.add(matched)
            if matched in OBSOLETE_STANDARD_PATHS:
                issues.append({
                    "file": str(workflow_path),
                    "line": index + 1,
                    "type": "obsolete_standard_path",
                    "path": matched,
                    "message": f"workflow 使用了过期标准路径: {matched}",
                })
                continue
            if not (
                matched.startswith("dev_docs/plans/")
                or matched.startswith("dev_docs/memos/")
                or matched.startswith("dev_docs/review/")
            ):
                continue
            if matched not in spec_paths and not _context_is_optional(lines, index):
                issues.append({
                    "file": str(workflow_path),
                    "line": index + 1,
                    "type": "workflow_path_drift",
                    "path": matched,
                    "message": f"workflow 使用了未被 spec 收录的标准路径: {matched}",
                })
    return {"checked": 1, "seen_standard_paths": sorted(seen_standard_paths), "issues": issues}


def render_text(payload):
    summary = payload["summary"]
    status = "PASS" if summary["passed"] else "FAIL"
    lines = [f"{status}: framework_contract_checker"]
    for check_name, result in payload["checks"].items():
        lines.append(f"- {check_name}: {len(result['issues'])} issue(s)")
        for issue in result["issues"]:
            location = issue.get("file", "")
            if issue.get("line"):
                location = f"{location}:{issue['line']}"
            lines.append(f"  - {issue['type']} {location} {issue.get('message', '')}".rstrip())
    return "\n".join(lines) + "\n"


def main():
    parser = argparse.ArgumentParser(description="检查 AICC 框架契约与 workflow 漂移")
    parser.add_argument("--self-check", action="store_true", help="检查当前仓库关键模板与 workflow")
    parser.add_argument("--check-template", help="检查指定模板或文档是否满足主文档契约")
    parser.add_argument("--check-workflow", help="检查指定 workflow 是否与 spec 路径一致")
    parser.add_argument("--contract", default=str(CONTRACTS_DIR / "main_doc_contract.yaml"), help="主文档契约文件路径")
    parser.add_argument("--spec", default=str(DEFAULT_SPEC), help="规范文档路径，用于 workflow 漂移检查")
    parser.add_argument("--format", choices=["json", "text"], default="json", help="输出格式")
    args = parser.parse_args()

    checks = {}

    if args.check_template:
        checks["template"] = check_template(Path(args.check_template), Path(args.contract))

    if args.check_workflow:
        checks["workflow"] = check_workflow(Path(args.check_workflow), Path(args.spec))

    if args.self_check:
        checks["template"] = check_template(DEFAULT_TEMPLATE, Path(args.contract))
        checks["spec_paths"] = check_spec_paths(Path(args.spec))
        workflow_issues = []
        seen_standard_paths = set()
        for workflow_path in DEFAULT_WORKFLOWS:
            workflow_result = check_workflow(workflow_path, Path(args.spec))
            workflow_issues.extend(workflow_result["issues"])
            seen_standard_paths.update(workflow_result.get("seen_standard_paths", []))
        checks["workflow"] = {
            "checked": len(DEFAULT_WORKFLOWS),
            "seen_standard_paths": sorted(seen_standard_paths),
            "issues": workflow_issues,
        }

    if not checks:
        parser.print_help()
        return 2

    total_issues = sum(len(result["issues"]) for result in checks.values())
    payload = {
        "summary": {
            "passed": total_issues == 0,
            "issues": total_issues,
        },
        "checks": checks,
    }

    if args.format == "text":
        sys.stdout.write(render_text(payload))
    else:
        json.dump(payload, sys.stdout, ensure_ascii=False, indent=2)
        sys.stdout.write("\n")

    return 0 if total_issues == 0 else 1


if __name__ == "__main__":
    sys.exit(main())

File: tools/js/framework_contract_checker.js
#!/usr/bin/env node
/**
 * 框架契约检查器
 *
 * 最小能力：
 * - 检查主文档模板是否满足 main_doc_contract.yaml 必需章节
 * - 检查 workflow 是否引入了未被 spec 收录的标准产物路径
 * - 提供框架自检模式
 */

const fs = require('fs');
const path = require('path');

const ROOT = path.resolve(__dirname, '..', '..');
const CONTRACTS_DIR = path.join(ROOT, 'core', 'contracts');
const DEFAULT_TEMPLATE = path.join(ROOT, 'templates', 'AI_Coding_Context_TEMPLATE.md');
const DEFAULT_SPEC = path.join(ROOT, 'core', 'framework_spec.md');
const DEFAULT_WORKFLOWS = [
  path.join(ROOT, 'workflows', 'generation_workflow.md'),
  path.join(ROOT, 'workflows', 'path_a_first_generation.md'),
];
const OBSOLETE_STANDARD_PATHS = new Set([
  'dev_docs/plans/features/',
  'dev_docs/plans/bugfixes/',
]);

function parseArgs() {
  const argv = process.argv.slice(2);
  const out = {
    contract: path.join(CONTRACTS_DIR, 'main_doc_contract.yaml'),
    spec: DEFAULT_SPEC,
    format: 'json',
  };
  for (let i = 0; i < argv.length; i++) {
    const arg = argv[i];
    const next = argv[i + 1];
    switch (arg) {
      case '--self-check': out.selfCheck = true; break;
      case '--check-template': out.checkTemplate = next; i++; break;
      case '--check-workflow': out.checkWorkflow = next; i++; break;
      case '--contract': out.contract = next; i++; break;
      case '--spec': out.spec = next; i++; break;
      case '--format': out.format = next; i++; break;
      case '-h':
      case '--help': out.help = true; break;
    }
  }
  return out;
}

function loadRequiredTitles(contractPath) {
  return fs.readFileSync(contractPath, 'utf8')
    .split('\n')
    .map((line) => {
      const match = line.match(/^\s*title:\s*"?(.*?)"?\s*$/);
      return match ? match[1] : null;
    })
    .filter(Boolean);
}

function extractH2Titles(text) {
  const titles = [];
  const regex = /^##\s+(.+?)\s*$/gm;
  let match;
  while ((match = regex.exec(text)) !== null) {
    titles.push(match[1].trim());
  }
  return titles;
}

function checkTemplate(templatePath, contractPath) {
  const requiredTitles = loadRequiredTitles(contractPath);
  const actualTitles = extractH2Titles(fs.readFileSync(templatePath, 'utf8'));
  const issues = [];
  for (const title of requiredTitles) {
    if (!actualTitles.includes(title)) {
      issues.push({
        file: templatePath,
        type: 'missing_required_section',
        section: title,
        message: `缺少必需章节: ${title}`,
      });
    }
  }
  return { checked: 1, issues };
}

function extractStandardPaths(specPath) {
  const text = fs.readFileSync(specPath, 'utf8');
  const paths = new Set();
  text.split('\n').forEach((line) => {
    if (line.includes('❌')) return;
    Array.from(line.matchAll(/`(dev_docs\/[^`]+)`/g), (match) => match[1])
      .forEach((targetPath) => paths.add(targetPath));
  });
  return paths;
}

function normalizeStandardPath(targetPath) {
  if (targetPath === 'dev_docs/plans/README.md') return 'dev_docs/plans/';
  if (targetPath.startsWith('dev_docs/plans/active/')) return 'dev_docs/plans/active/';
  if (targetPath.startsWith('dev_docs/plans/done/')) return 'dev_docs/plans/done/';
  if (targetPath.startsWith('dev_docs/plans/archive/')) return 'dev_docs/plans/archive/';
  if (targetPath.startsWith('dev_docs/memos/')) return 'dev_docs/memos/';
  if (targetPath.startsWith('dev_docs/knowledge/')) return 'dev_docs/knowledge/';
  if (targetPath.startsWith('dev_docs/rules/combined/')) return 'dev_docs/rules/combined/AI_RULES.md';
  return targetPath;
}

function extractUserDocPaths(text) {
  return Array.from(text.matchAll(/dev_docs\/[\w./-]+/g), (match) => normalizeStandardPath(match[0]));
}

function checkSpecPaths(specPath) {
  const standardPaths = Array.from(extractStandardPaths(specPath)).sort();
  const issues = [];
  for (const obsoletePath of Array.from(OBSOLETE_STANDARD_PATHS).sort()) {
    if (standardPaths.includes(obsoletePath)) {
      issues.push({
        file: specPath,
        type: 'obsolete_standard_path',
        path: obsoletePath,
        message: `spec 仍包含过期标准路径: ${obsoletePath}`,
      });
    }
  }
  return { checked: 1, standard_paths: standardPaths, issues };
}

function contextIsOptional(lines, index) {
  let start = 0;
  for (let cursor = index; cursor >= 0; cursor--) {
    if (/^#{2,6}\s+/.test(lines[cursor])) {
      start = cursor;
      break;
    }
  }
  const end = Math.min(lines.length, index + 4);
  const context = lines.slice(start, end).join('\n');
  const markers = ['[OPTIONAL]', '不属于', '只有在用户明确需要', '可选', '不是必需'];
  return markers.some((marker) => context.includes(marker));
}

function checkWorkflow(workflowPath, specPath) {
  const specPaths = extractStandardPaths(specPath);
  const lines = fs.readFileSync(workflowPath, 'utf8').split('\n');
  const issues = [];
  const seenStandardPaths = new Set();
  lines.forEach((line, index) => {
    for (const driftPath of extractUserDocPaths(line)) {
      if (specPaths.has(driftPath)) {
        seenStandardPaths.add(driftPath);
      }
      if (OBSOLETE_STANDARD_PATHS.has(driftPath)) {
        issues.push({
          file: workflowPath,
          line: index + 1,
          type: 'obsolete_standard_path',
          path: driftPath,
          message: `workflow 使用了过期标准路径: ${driftPath}`,
        });
        continue;
      }
      if (
        !driftPath.startsWith('dev_docs/plans/')
        && !driftPath.startsWith('dev_docs/memos/')
        && !driftPath.startsWith('dev_docs/review/')
      ) {
        continue;
      }
      if (!specPaths.has(driftPath) && !contextIsOptional(lines, index)) {
        issues.push({
          file: workflowPath,
          line: index + 1,
          type: 'workflow_path_drift',
          path: driftPath,
          message: `workflow 使用了未被 spec 收录的标准路径: ${driftPath}`,
        });
      }
    }
  });
  return { checked: 1, seen_standard_paths: Array.from(seenStandardPaths).sort(), issues };
}

function renderText(payload) {
  const status = payload.summary.passed ? 'PASS' : 'FAIL';
  const lines = [`${status}: framework_contract_checker`];
  Object.entries(payload.checks).forEach(([name, result]) => {
    lines.push(`- ${name}: ${result.issues.length} issue(s)`);
    result.issues.forEach((issue) => {
      const location = issue.line ? `${issue.file}:${issue.line}` : issue.file;
      lines.push(`  - ${issue.type} ${location} ${issue.message || ''}`.trimEnd());
    });
  });
  return `${lines.join('\n')}\n`;
}

function main() {
  const args = parseArgs();
  if (args.help) {
    process.stdout.write(
      'Usage: framework_contract_checker [--self-check] [--check-template FILE] [--check-workflow FILE] [--contract YAML] [--spec FILE] [--format json|text]\n'
    );
    return 0;
  }

  const checks = {};

  if (args.checkTemplate) {
    checks.template = checkTemplate(args.checkTemplate, args.contract);
  }

  if (args.checkWorkflow) {
    checks.workflow = checkWorkflow(args.checkWorkflow, args.spec);
  }

  if (args.selfCheck) {
    checks.template = checkTemplate(DEFAULT_TEMPLATE, args.contract);
    checks.spec_paths = checkSpecPaths(args.spec);
    const issues = [];
    const seenStandardPaths = new Set();
    DEFAULT_WORKFLOWS.forEach((workflowPath) => {
      const workflowResult = checkWorkflow(workflowPath, args.spec);
      issues.push(...workflowResult.issues);
      workflowResult.seen_standard_paths.forEach((seenPath) => seenStandardPaths.add(seenPath));
    });
    checks.workflow = {
      checked: DEFAULT_WORKFLOWS.length,
      seen_standard_paths: Array.from(seenStandardPaths).sort(),
      issues,
    };
  }

  if (Object.keys(checks).length === 0) {
    process.stdout.write(
      'Usage: framework_contract_checker [--self-check] [--check-template FILE] [--check-workflow FILE] [--contract YAML] [--spec FILE] [--format json|text]\n'
    );
    return 2;
  }

  const totalIssues = Object.values(checks).reduce((sum, result) => sum + result.issues.length, 0);
  const payload = {
    summary: {
      passed: totalIssues === 0,
      issues: totalIssues,
    },
    checks,
  };

  if (args.format === 'text') {
    process.stdout.write(renderText(payload));
  } else {
    process.stdout.write(`${JSON.stringify(payload, null, 2)}\n`);
  }

  return totalIssues === 0 ? 0 : 1;
}

process.exit(main());
